### ISSUE TITLE
Proposal: split stactools into `stactools` and `stactools-plugins`

### ISSUE THREAD

**Issue by gadomski** [stac-utils/stactools#102]
The current **stactools** packaging story is a bit broken, as noted in #98. This repository is trying to be two things at once:

1. A "swiss army knife" of capabilities that a STAC user might want to use via the command line or via an API, e.g. moving, copying, and merging.
2. Dataset-specific conversion utilities to create STAC items from assets (and maybe modify those assets).

I'm opening this issue to lay out the argument for splitting **stactools** into two separate repositories: **stactools** and **stactools-plugins**, both living in the **stac-utils** organiziation. **stactools** would contain `stactools.core`, `stactools.cli`, and `stactools.browse`. The rest of the dataset-specific subpackages would move to **stactools-plugins**. These would be separately installable via pip (e.g. `pip install stactools-planet`) and would integrate with the `stactools` API via [namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/) and with the `stactools.cli` via the (currently used) plugin mechanism. This isn't too dissimilar to how things work right now, but would allow the core **stactools** code to move at its own development cadence relative to the dataset-specific tools.

Other related discussions: #95, https://github.com/stac-utils/stactools/pull/84#discussion_r622361095#100 

Pinging in @scottyhq and @kylebarron as they've been involved in some background conversations that led to this proposal.

**Comment by scottyhq** [comment]
Thanks for opening this @gadomski and laying things out very clearly. It does seem like decoupling things in a way such that data-specific conversion utilities are independent and separate will facilitate adding new data sources. I imagine the list of subfolders could grow quite rapidly as STAC gains popularity and more public data is hosted by cloud providers.

I don't have strong opinions on how things are setup, so long as `pip install -e stactools` works for *development* of a single subpackage and subpackages can be installed independently `pip install stactools[sentinel1]` (or some variant).

I did also want to point out an alternative organization that the [intake library](https://intake.readthedocs.io/en/latest/making-plugins.html#entrypoints) uses. There are lots of parallels between the intake catalog system and what stactools is doing. Python package entrypoints are another approach for organizing a core library with plugins that live in their own repositories rather than under subfolders as subpackages. see also https://blog.danallan.com/posts/2019-07-24-use-entrypoints-more/

**Comment by lossyrob** [comment]
Thank you @gadomski for driving this! I'm in favor of this split, so +1 from me.

Interesting about the entrypoints technique - I hadn't seen that before. When I started stactools I had cribbed the namespace package technique from rastervision - credit to @lewfish! (Tagging for credit, but also in case this might be interesting to rastervision's subpackage approach, though I'm not sure how that's changed since I last looked). The points in the blog post you linked to are convincing, and if entrypoints are a relatively well worn mechanism to do this sort of plugin discovery, moving to it would make sense to me.

**Comment by gadomski** [comment]
I've opened a (draft for now) PR #111 to show my current thinking on this issue. As I was splitting out subpackages into a separate plugins repository, I felt like I was doing lot of work for not that much benefit; most of the subpackages have light-to-no extra dependencies beyond what's required in `stactools.core`, so all of the heavy lifting to keep the dependency profile low seemed extra.

**Comment by JamesTownend** [comment]
As a someone going through the contribution process, the proposed structure looks friendlier. However, is there potential for the dependencies to get out of hand if many new sub-packages are included? Could we leverage Github's template repositories to facilitate both ease of sub-package contribution and separation of dependencies?

**Comment by gadomski** [comment]
I agree the scaling could become a problem down the road if we grow to have a huge number of subpackages. The main problem I forsee with having a large number of dataset subpackges in their own repos is those subpackage repos falling out-of-sync with the main **stactools** API. If `stactools_foo` wants `stactools_core==1.2` but `stactools_bar` needs `stactools_core==2`, we've got a problem. By keeping our "core" dataset subpackages in one (or maybe two, see my original plugin repository idea), we make it easier to ensure these core subpackages stay up-to-date using one or two CI hooks.

So far, it seems like most dataset subpackages don't have too many additional Python dependencies, and what additional dependencies there are (e.g. `boto3`) aren't too hard to manage with the `extras_require` mechanism. I don't see a big issue with having many (20, 30, more) dataset subpackages in the main repo, since in general these subpackages are relatively static (once they're built, they hopefully just keep working w/ minor tweaks to keep them up to the latest `stactools_core` API).

**Comment by JamesTownend** [comment]
In that case, would it be worth creating a subpackage template folder? The template could include suggested CLI commands to help standardise across subpackages. I'm happy to create it if you think it's worthwhile.

**Comment by gadomski** [comment]
Yeah, take a look at https://github.com/stac-utils/stactools/pull/111 see if that makes sense, and if you think a template would be necessary/helpful. I find templates a double-edged sword at times -- they can be helpful for standing up new subpackages, but they can get out-of-date. If we could have a template dataset subpackage that could be included in CI to ensure it's relatively fresh then 👍🏽.

**Comment by cuttlefish** [comment]
I took a quick look at #111 and it seems reasonable to me. Separating the dataset specific code from the main modules makes sense to me. After a bit of discussion I'm somewhat leaning towards individual repos. I think the risk of some unloved subpackages lagging behind is outweighed by allowing the rest to move forward. It would also make it easier to write custom data converters for private datasets. But either way the new packaging structure should support both. Is #111 still being blocked by a couple of outstanding PRs?

**Comment by gadomski** [comment]
Yeah I'm keeping #111 as a draft during these discussions. If we do decide to split out dataset-specific packages to their own repos (which is the way things seem to be going), we can use #111 as the starting point and just remove the dataset stuff from it.

**Comment by gadomski** [comment]
Updates:

- I've updated #111 to match our current thinking on the reorganization (i.e. subpackages will be split out to their own repos).
- I've got a proof-of-concept subpackage template repository here: https://github.com/gadomski/stactools-subpackage-template. 
- I've created a script that extracts subpackages from the stactools repo (including their Github repository) here: https://gist.github.com/gadomski/a0ab8d18f56ff34454d2ed890d234534
- I've got a proof-of-concept of a dataset subpackage here (created using the above script): https://github.com/gadomski/stactools-threedep

Some thoughts/proposals:

- If you want a certain dataset's subpackage, you can just do a `pip install stactools_planet` and get what you need. The subpackages will pin whatever version of `stactools` they want, which can (and will) vary from subpackage to subpackage. This will be helpful e.g. in downstream dataset-specific Docker containers.
- We won't specify subpackages as dependencies or "extras_require" in `stactools`, but we will track their build status via a (non-blocking) Github actions workflow. This will allow us to monitor the freshness of key subpackages while still allowing new development not to be blocked by atrophying other packages (looking at you, landsat).
- We will probably want to create a docker image and conda package for `stactools`, so subpackages can depend on those rather than duplicating some code (e.g. conda's `environment.yml` file and the dockerfiles).

**Comment by vincentsarago** [comment]
👋 Hi, and sorry for jumping quite late on this issue. 

In TiTiler and stac-fastapi (https://github.com/stac-utils/stac-fastapi/pull/125) we choose to use `namespace` packages to add the possibility to add more extension without breaking all other package. Having another org or repo for dataset might be quite confusing for user/dev and I think it would be nice to keep everything here.

IMO it would be nice to keep a standard way of subpackaging through the stactool org.

Feel free to ignore this comment because I won't have time to spend on code or review in the next weeks 🙉

**Comment by gadomski** [comment]
👋🏽 no worries. Yes, we did consider the tradeoffs between keeping things in one repository or splitting them out (see https://github.com/stac-utils/stactools/issues/102#issuecomment-839122602 and https://github.com/stac-utils/stactools/issues/102#issuecomment-853006964 for some of the waffling). In the end, we decided that separate repos would be useful for the following two main reasons:

1. A template repository provides an easier on-ramp for developers new to stactools to contribute a new dataset package.
2. By splitting the packages into their own repos, we prevent bitrot in one package (e.g. landsat, see https://github.com/stac-utils/stactools/pull/113#issue-650133231) from blocking progress in other datasets or in `stactools.core`.

Point 2 in particular is important for `stactools`, since we're going to accumulate a large number of dataset packages that might be used heavily for a short period of time (e.g. to ingest into some new cloud provider), and then will be mostly dormant. We don't want these packages blocking development by, e.g., failing CI; at the same time, we don't want to remove or disable them altogether even if they're dormant because they might be useful later. Separate repositories seemed like the best way to allow each of these dataset packages to proceed at their own cadence.

From a Python API perspective, we're still using namespace packages in our general architecture. For the dataset-agnotic portions of `stactools`, `stactools.cli` and `stactools.core`, we are following the same model as `stac-fastapi`. So while there's some divergence, we're not _too_ far off of the trail IMO. Appreciate the feedback and please jump in as we go along with any other suggestions or comments! 👍🏽
